### PR TITLE
Add system and channel support with series compat.

### PR DESCRIPTION
**Pull request by hpidcock**
- Adds systems support
- Systems can be serialized to and from a series.
- Channel support is repurposed from snapd.

### CODE REVIEW

File: builtin.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+package systems
+
+import (
+	"github.com/juju/collections/set"
+
+	"github.com/juju/systems/channel"
+)
+
+const (
```
Comment (wallyworld) [inline builtin.go:16]: add a comment for the block of constants

File: builtin.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	CentOS       = "centos"
+	Windows      = "windows"
+	OSX          = "osx"
+	OpenSUSE     = "opensuse"
+	GenericLinux = "genericlinux"
+)
+
+var validOS = set.NewStrings(Ubuntu, CentOS, Windows, OSX, OpenSUSE, GenericLinux)
+
+var seriesToSystem = map[string]System{
```
Comment (wallyworld) [inline builtin.go:30]: Do we need all these old ones? Maybe just trusty, xenial, bionic, focal, eoan, groovy?

File: channel/channel.go [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+	Risk   Risk   `json:"risk"`
+	Branch string `json:"branch,omitempty"`
+}
+
+func isSlash(r rune) bool { return r == '/' }
+
+// TODO: currently there's some overlap between the toplevel Full, and
+//       methods Clean, String, and Full. Needs further refactoring.
+
+func Full(s string) (string, error) {
```
Comment (wallyworld) [inline channel/channel.go:57]: comment

File: channel/channel.go [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+	if s == "" {
+		return "", nil
+	}
+	components := strings.FieldsFunc(s, isSlash)
```
Comment (wallyworld) [inline channel/channel.go:61]: with only 1 delimiter, could use Split() ?

File: channel/channel.go [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+	"github.com/juju/collections/set"
+)
+
+// Empty channel
+var Empty Channel
+
+// Risk type
+type Risk string
+
+const (
```
Comment (wallyworld) [inline channel/channel.go:21]: comment the block

File: channel/channel.go [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+		fallthrough
+	case 3:
+		return strings.Join(components, "/"), nil
+	default:
+		return "", errors.New("invalid channel")
+	}
+}
+
+// ParseVerbatim parses a string representing a store channel and
+// includes the given architecture, if architecture is "" the system
```
Comment (wallyworld) [inline channel/channel.go]: i don't see "the given architecture"

File: channel/channel.go [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+	}
+	return c
+}
+
+// Clean returns a Channel with a normalized track, risk and name.
+func (c Channel) Clean() Channel {
+	track := c.Track
+	risk := c.Risk
+
+	if track == "latest" {
```
Comment (wallyworld) [inline channel/channel.go:158]: const for "latest" ?
maybe overkill

File: channel/channel.go [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+	if cm.Track {
+		matching = append(matching, "track")
+	}
+	if cm.Risk {
+		matching = append(matching, "risk")
+	}
+	return strings.Join(matching, ":")
+}
+
+// Match returns a Match of which fields among architecture,track,risk match between c and c1 store channels, risk is matched taking channel inheritance into account and considering c the requested channel.
```
Comment (wallyworld) [inline channel/channel.go:237]: where's architecture?

File: channel/channel.go [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+	}
+	p := strings.Split(newChannel, "/")
+	if channelRisks.Contains(p[0]) && ch.Track != "" {
+		// risk/branch inherits the track if any
+		return ch.Track + "/" + newChannel, nil
+	}
+	return newChannel, nil
+}
+
+var ErrPinnedTrackSwitch = errors.New("cannot switch pinned track")
```
Comment (wallyworld) [inline channel/channel.go:271]: comment

File: system.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+		if s.Channel == channel.Empty {
+			return errors.NotValidf("missing channel")
+		}
+	} else if s.Resource != "" {
+		if s.Channel != channel.Empty {
+			return errors.NotValidf("channel cannot be specified with resource")
+		}
+	} else {
+		return errors.NotValidf("one of os or resource must be specified")
+	}
```
Comment (wallyworld) [inline system.go]: Maybe do this check first,
```
if s.OS == "" && s.Resource == "" {
    return errors.NotValidf("one of os or resource must be specified")	
}
```

then the rest....

Maybe makes it easier to read?

File: system.go [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+	matchedCharacters := 0
+	system := System{}
+	for _, v := range matches {
+		matchedCharacters += len(v[0])
+		key := v[1]
+		value := v[2]
+		switch key {
+		case "os":
+			system.OS = value
+		case "channel":
```
Comment (wallyworld) [inline system.go:99]: I don't see channel in the regexp?

File: system.go [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+	}
+	if s.Resource != "" {
+		str += fmt.Sprintf("#resource=%s", s.Resource)
+	}
+	return str
+}
+
+// regex to match k=v from system series strings in the form
+// "system#os=ubuntu#version=18.04#resource=imagename"
+var systemSeriesRegex = regexp.MustCompile(`#(os|version|resource)=([^#]+)`)
```
Comment (wallyworld) [inline system.go]: Did you mean channel not version?

File: system_test.go [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+	"github.com/juju/systems"
+	"github.com/juju/systems/channel"
+)
+
+type systemSuite struct {
+	testing.CleanupSuite
+}
+
+var _ = gc.Suite(&systemSuite{})
```
Comment (wallyworld) [inline system_test.go:21]: It seems some tests are missing to catch the issues raised above

eg we have "system#os=ubuntu" but there's no test for a full system string

**Review by wallyworld — APPROVED** [review]
Needs a couple of fixes I think before landing

File: builtin.go (hunk 2 above)
Comment (hpidcock) [reply builtin.go:30]: I want to make sure that new charms/updated charms for IAAS can still use old series.

**Pull request merged**